controllers/pagerdutyintegration/serviceorchestration_handler.go

```diff
@@ -0,0 +1,126 @@
+package pagerdutyintegration
+
+import (
+	"context"
+	"fmt"
+	hivev1 "github.com/openshift/hive/apis/hive/v1"
+	pagerdutyv1alpha1 "github.com/openshift/pagerduty-operator/api/v1alpha1"
+	"github.com/openshift/pagerduty-operator/config"
+	"github.com/openshift/pagerduty-operator/pkg/localmetrics"
+	pd "github.com/openshift/pagerduty-operator/pkg/pagerduty"
+	"github.com/openshift/pagerduty-operator/pkg/utils"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	ServiceOrchestrationDataName = "service-orchestration.json"
+)
+
+// handleServiceOrchestration enables and applies the service orchestration rule to the PD service if it is enabled in PDI
+func (r *PagerDutyIntegrationReconciler) handleServiceOrchestration(pdclient pd.Client, pdi *pagerdutyv1alpha1.PagerDutyIntegration, cd *hivev1.ClusterDeployment) error {
+	var (
+		// clusterConfigmapName is the name of the ConfigMap containing the
+		// SERVICE_ID and INTEGRATION_ID
+		clusterConfigmapName string = config.Name(pdi.Spec.ServicePrefix, cd.Name, config.ConfigMapSuffix)
+
+		// orchestrationConfigmapName is the name of the configmap containing the
+		// service orchestration rules
+		orchestrationConfigmapName string = pdi.Spec.ServiceOrchestration.RuleConfigConfigMapRef.Name
+
+		fakeClusterDeploymentAnnotation = "managed.openshift.com/fake"
+	)
+
+	if !cd.Spec.Installed {
+		// Cluster isn't installed yet, return
+		return nil
+	}
+
+	val, ok := cd.Annotations[fakeClusterDeploymentAnnotation]
+	if ok && val == "true" {
+		r.reqLogger.Info("Fake cluster identified: " + cd.Spec.ClusterName + ". Skipping reconcile.")
+		return nil
+	}
+
+	clusterID := utils.GetClusterID(cd)
+	pdData, err := pd.NewData(pdi, clusterID, cd.Spec.BaseDomain)
+	if err != nil {
+		return err
+	}
+
+	clusterConfigMap := &corev1.ConfigMap{}
+
+	err = r.Client.Get(context.TODO(), types.NamespacedName{
+		Namespace: cd.Namespace, Name: clusterConfigmapName}, clusterConfigMap)
+	if err != nil {
+		return err
+	}
+
+	// load configuration
+	err = pdData.ParseClusterConfig(r.Client, cd.Namespace, clusterConfigmapName)
+	if err != nil {
+		return err
+	}
+
+	pdiSelector, err := metav1.LabelSelectorAsSelector(&pdi.Spec.ClusterDeploymentSelector)
+	if err != nil {
+		return err
+	}
+
+	if pdData.ServiceID == "" {
+		// PD service has not been created, skip the service orchestration steps
+		return nil
+	}
+	if pdData.ServiceOrchestrationState != "active" {
+		r.reqLogger.Info("enabling the service orchestration")
+		err := pdclient.ToggleServiceOrchestration(pdData, true)
```

Comment:
Wouldn't this enable ServiceOrchestration for every single PagerDuty service? (Is that intended? I may have forgotten)

Reply:
In pagerdutyintegration_controller.go, this whole `handleServiceOrchestration()` function will be called only when `if pdi.Spec.ServiceOrchestration.Enabled`.
So the goal here is going to enable the service orchestration for all the osd/rosa clusters' PD service which match the selectors defined in the PagerdutyIntegration `osd`.

Reply:
Right, thanks for clarifying - that makes much more sense!